File: src/main/python/AoC2015_06.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 6
#

from __future__ import annotations

import sys
from enum import Enum
from enum import auto
from enum import unique
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import Cell

TEST1 = "turn on 0,0 through 999,999"
TEST2 = "toggle 0,0 through 999,0"
TEST3 = "turn off 499,499 through 500,500"
TEST4 = "turn on 0,0 through 0,0"
TEST5 = "toggle 0,0 through 999,999"


@unique
class Mode(Enum):
    MODE_1 = auto()
    MODE_2 = auto()


@unique
class Action(Enum):
    TURN_ON = auto()
    TOGGLE = auto()
    TURN_OFF = auto()

    @classmethod
    def from_string(cls, action: str) -> Action:
        match action:
            case "turn_on":
                return Action.TURN_ON
            case "turn_off":
                return Action.TURN_OFF
            case _:
                return Action.TOGGLE

    def apply(
        self, lights: list[list[int]], start: Cell, end: Cell, mode: Mode
    ) -> None:
        match self:
            case Action.TURN_ON:
                match mode:
                    case Mode.MODE_1:
                        for r in range(start.row, end.row + 1):
                            for c in range(start.col, end.col + 1):
                                lights[r][c] = 1
                    case Mode.MODE_2:
                        for r in range(start.row, end.row + 1):
                            for c in range(start.col, end.col + 1):
                                lights[r][c] += 1
            case Action.TURN_OFF:
                match mode:
                    case Mode.MODE_1:
                        for r in range(start.row, end.row + 1):
                            for c in range(start.col, end.col + 1):
                                lights[r][c] = 0
                    case Mode.MODE_2:
                        for r in range(start.row, end.row + 1):
                            for c in range(start.col, end.col + 1):
                                lights[r][c] = max(lights[r][c] - 1, 0)
            case Action.TOGGLE:
                match mode:
                    case Mode.MODE_1:
                        for r in range(start.row, end.row + 1):
                            for c in range(start.col, end.col + 1):
                                lights[r][c] = 0 if lights[r][c] == 1 else 1
                    case Mode.MODE_2:
                        for r in range(start.row, end.row + 1):
                            for c in range(start.col, end.col + 1):
                                lights[r][c] += 2


class Instruction(NamedTuple):
    action: Action
    start: Cell
    end: Cell

    @classmethod
    def from_input(cls, input_: str) -> Instruction:
        splits = input_.replace("turn ", "turn_").split(" through ")
        action_s, start_s = splits[0].split(" ")
        start = Cell(*list(map(int, start_s.split(",")))[:2])
        end = Cell(*list(map(int, splits[1].split(",")))[:2])
        return Instruction(Action.from_string(action_s), start, end)


Input = list[Instruction]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [Instruction.from_input(line) for line in input_data]

    def solve(self, instructions: Input, mode: Mode) -> int:
        lights = [[0 for _ in range(1000)] for _ in range(1000)]
        for instruction in instructions:
            instruction.action.apply(
                lights, instruction.start, instruction.end, mode
            )
        return sum(lights[r][c] for r in range(1000) for c in range(1000))

    def part_1(self, instructions: Input) -> Output1:
        return self.solve(instructions, Mode.MODE_1)

    def part_2(self, instructions: Input) -> Output2:
        return self.solve(instructions, Mode.MODE_2)

    @aoc_samples(
        (
            ("part_1", TEST1, 1_000_000),
            ("part_1", TEST2, 1000),
            ("part_1", TEST3, 0),
            ("part_2", TEST4, 1),
            ("part_2", TEST5, 2_000_000),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2015, 6)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2015_06.java
import static java.util.stream.Collectors.toList;

import java.util.Arrays;
import java.util.List;
import java.util.function.IntFunction;

import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

// TODO: better -> https://www.reddit.com/r/adventofcode/comments/3vmltn/day_6_solutions/cxozgap/
public final class AoC2015_06
            extends SolutionBase<List<AoC2015_06.Instruction>, Integer, Integer> {
    
    private AoC2015_06(final boolean debug) {
        super(debug);
    }

    public static AoC2015_06 create() {
        return new AoC2015_06(false);
    }

    public static AoC2015_06 createDebug() {
        return new AoC2015_06(true);
    }
    
    @Override
    protected List<Instruction> parseInput(final List<String> inputs) {
        return inputs.stream().map(Instruction::fromInput).collect(toList());
    }

    private int solve(final List<Instruction> instructions, final Mode mode) {
        final int[] lights = new int[1_000_000];
        for (final Instruction instruction : instructions) {
            instruction.action.apply(
                    lights, instruction.start, instruction.end, mode);
        }
        return Arrays.stream(lights).sum();
    }

    @Override
    public Integer solvePart1(final List<Instruction> instructions) {
        return solve(instructions, Mode.MODE_1);
    }

    @Override
    public Integer solvePart2(final List<Instruction> instructions) {
        return solve(instructions, Mode.MODE_2);
    }
    
    enum Mode { MODE_1, MODE_2 }

    record Instruction(Action action, Cell start, Cell end) {
        public static Instruction fromInput(final String input) {
            final String s = input.replace("turn ", "turn_");
            final StringSplit splits = StringOps.splitOnce(s, " through ");
            final StringSplit actionAndStartSplits
                    = StringOps.splitOnce(splits.left(), " ");
            return new Instruction(
                Action.fromString(actionAndStartSplits.left()),
                Cell.fromString(actionAndStartSplits.right()),
                Cell.fromString(splits.right()));
        }
        
        enum Action {
            TURN_ON, TURN_OFF, TOGGLE;

            public static Action fromString(final String s) {
                return switch (s) {
                    case "turn_on" -> TURN_ON;
                    case "turn_off" -> TURN_OFF;
                    default -> TOGGLE;
                };
            }

            public void apply(
                    final int[] grid,
                    final Cell start,
                    final Cell end,
                    final Mode mode
            ) {
                final IntFunction<Integer> f = switch (this) {
                    case TURN_ON -> switch (mode) {
                            case MODE_1: yield v -> 1;
                            case MODE_2: yield v -> v + 1;
                        };
                    case TURN_OFF -> switch (mode) {
                            case MODE_1: yield v -> 0;
                            case MODE_2: yield v -> Math.max(v - 1, 0);
                        };
                    case TOGGLE -> switch (mode) {
                            case MODE_1: yield v -> v == 1 ? 0 : 1;
                            case MODE_2: yield v -> v + 2;
                        };
                };
                final int rStart = start.getRow() * 1_000;
                final int rEnd = end.getRow() * 1_000;
                for (int r = rStart; r <= rEnd; r += 1_000) {
                    for (int idx = r + start.getCol(); idx <= r + end.getCol(); idx++) {
                        grid[idx] = f.apply(grid[idx]);
                    }
                }
            } //apply
        } // Action
    } // Instruction

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "1000000"),
        @Sample(method = "part1", input = TEST2, expected = "1000"),
        @Sample(method = "part1", input = TEST3, expected = "0"),
        @Sample(method = "part2", input = TEST4, expected = "1"),
        @Sample(method = "part2", input = TEST5, expected = "2000000"),
    })
    public void samples() {
    }

    public static void main(final String[] args) throws Exception {
        AoC2015_06.create().run();
    }

    private static final String TEST1 = "turn on 0,0 through 999,999";
    private static final String TEST2 = "toggle 0,0 through 999,0";
    private static final String TEST3 = "turn off 499,499 through 500,500";
    private static final String TEST4 = "turn on 0,0 through 0,0";
    private static final String TEST5 = "toggle 0,0 through 999,999";
}